presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduMetadata.java

```diff
@@ -450,8 +453,29 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
                 handle.getTable(clientSession),
                 handle.getConstraint(),
                 Optional.of(desiredColumns.build()),
-                handle.isDeleteHandle());
+                handle.isDeleteHandle(),
+                handle.getLimit());
 
         return Optional.of(new ProjectionApplicationResult<>(handle, projections, assignmentList.build()));
     }
+
+    @Override
+    public Optional<LimitApplicationResult<ConnectorTableHandle>> applyLimit(ConnectorSession session, ConnectorTableHandle table, long limit)
+    {
+        KuduTableHandle handle = (KuduTableHandle) table;
+
+        if (handle.getLimit().isPresent() && handle.getLimit().getAsLong() <= limit) {
+            return Optional.empty();
+        }
+
+        handle = new KuduTableHandle(
+                handle.getSchemaTableName(),
+                handle.getTable(clientSession),
+                handle.getConstraint(),
+                handle.getDesiredColumns(),
+                handle.isDeleteHandle(),
+                OptionalLong.of(limit));
+
+        return Optional.of(new LimitApplicationResult<>(handle, false));
```

Comment:
why is it not guaranteed?

Reply:
I think it creates limit for each partition so we can't guarantee a global limit but the limit is applied per partition